src/lib.rs

```diff
@@ -349,28 +349,43 @@ impl RouteRefresh {
     }
 
     /// Encode RouteRefresh to bytes
-    pub fn encode(&self, buf: &mut dyn Write) -> Result<(), Error> {
+    pub fn encode<W: Write>(&self, buf: &mut W) -> Result<(), Error> {
         buf.write_u16::<BigEndian>(self.afi as u16)?;
         buf.write_u8(self.subtype)?;
         buf.write_u8(self.safi as u8)
     }
 }
 
+/// An abstract way of getting a reference to a Capabilities struct.
```

Comment:
Would this make sense an an impl of [AsRef](https://doc.rust-lang.org/std/convert/trait.AsRef.html)?

Reply:
Yeah, I just tried this locally. After implementing `AsRef` for Capabilities (which looks identical to @TheBlueMatt 's changes above, we just need to update the `Reader` sig:

```
(struct/impl...) Reader<T, C>
where
    T: Read,
    C: AsRef<Capabilities>,
```

Reply:
Thanks for testing! I think this is the more preferred method then. It looks much cleaner as well and is more extensible in my opinion. Do you agree as well Matt?

Reply:
My initial patch used AsRef and then `impl AsRef<Capabilities> for Capabilities`, but this is super awkward. Sure, it works, but it suggests that Capabilities is a reference to a Capabilities, which is nonsense, so I went with a new trait. As for more motivation, maybe it helps to see where its used - https://github.com/TheBlueMatt/dnsseed-rust/blob/master/src/bgp_client.rs#L218. You'll note that its super, super awkward to *not* create a new Reader for each deserialize call in an async context, so you'd end up copying the Capabilities every time you receive a packet (ie more than once per message), which would have real performance overhead.

Reply:
Thanks for clarifying :) That makes sense!